feat: Add EtcdMetrics (DIS-297)

tzulingk opened the pull request:
#### Overview:

Add EtcdMetrics

#### Details:

Adding EtcdMetrics as a member to Etcd Client

```
pub struct Client {
    metrics: Arc<OnceLock<Arc<EtcdMetrics>>>,
}
```

Use OnceLock and lazy_initialization to initialize EtcdMetrics in DistributedRuntime::new()
```
if let Some(etcd_client) = &distributed_runtime.etcd_client {
    etcd_client.init_metrics(&distributed_runtime).await?;
}
```

Query the etcd server at time of init_metrics and set the values to EtcdMetrics during init_metrics().
```
        // Scan etcd for all known instance root paths to backfill metrics.
        // We must check all locations where leased instances might be registered.
        let prefixes_to_scan = vec![INSTANCE_ROOT_PATH];

        let mut total_initial_count = 0;
        let mut total_initial_bytes = 0;

        for prefix in prefixes_to_scan {
            let existing_instances = self.kv_get_prefix(prefix).await?;
            total_initial_count += existing_instances.len() as i64;
            total_initial_bytes += existing_instances
                .iter()
                .map(|kv| kv.value().len() as i64)
                .sum::<i64>();
        }

        metrics.etcd_block_total.add(total_initial_count);
        metrics.etcd_block_bytes_total.add(total_initial_bytes);
```


#### Where should the reviewer start?

lib/runtime/src/transports/etcd/metrics.rs: define EtcdMetrics
lib/runtime/src/transports/etcd.rs: initialize/set EtcdMetrics
lib/runtime/src/distributed.rs: call etcd_client.init_metrics()

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

DIS-297


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added real-time metrics tracking for etcd key-value operations, including monitoring the total number of keys and total bytes stored.
  * Metrics are now initialized automatically when the distributed runtime starts, providing immediate visibility into etcd storage usage.
* **Enhancements**
  * Metrics are updated live on key creation, updates, and deletions, ensuring accurate and up-to-date insights for users.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tzulingk committed:
Add EtcdMetrics

nnshah1 commented on lib/runtime/src/transports/etcd.rs:
```diff
@@ -32,11 +32,18 @@ pub use etcd_client::{ConnectOptions, KeyValue, LeaseClient};
+use metrics::*;
 pub use path::*;
 
+const NAMESPACE: &str = "system";
```
Question: why would this be a endpoint based metrics registry as opposed to a global / namespace based registry? I think its not tied to an endpoint - and these values are not technically namespace, component or endpoint as initialized in the system - might be missing something though

keivenchang replied:
It's my fault for not explaining the idea of endpoint clearly. I'll work this out with Tzu-ling.

tzulingk replied:
I updated init_metrics(). It's now using DistributedRuntime.create_intgauge()

grahamking committed:
fix: Simplify the types

- `metrics: Arc<OnceLock<Arc<EtcdMetrics>>>` to `metrics: Option<EtcdMetrics>`
- `EtcdMetrics`: remove both `Arc`.

The metrics creation functions should not be returning Arc. We can fix
that in later PR.

keivenchang commented:
Some things to consider:
- A DistributedRuntime object contains a single instance etcd_client
- All the children/grandchildren (Namespace, Component, Endpoint), will use the global drt().etcd_client to perform operations like kv_create, kv_put, etc
  - Note that these children also implement the *MetricsRegistry trait*
- The children call these etcd_client.kv_* functions like this: kv_*(etcd_path_with_lease_id(...), ...)
- The string etcd_path_with_lease_id(...) contains "instances/<namespace>/<component>/<endpoint>/<lease-id>" string
  - Not pretty solution: you can potentially parse that string to create Prometheus metric with that namespace/component/endpoint information
  - Prettier solution: you can potentially create new kv_*(MetricsRegistry, ...) functions, and pass in your Namespace/Component/Endpoint object into them -- then use these MetricsRegistry objects to create Prometheus metrics dynamically there, if needed.
  - @ryanolson what do you think? Any other more elegant way to do this?

keivenchang commented:
I just learned that @alec-flowers is in the middle of a refactor that will remove the etcd client, making this commit invalid. We will await for Alec's refactor.

tzulingk committed:
Use DistributedRuntime instead of Endpoint to create EtcdMetrics, since its a global metric

tzulingk commented:
<!-- linear:replyTo:3160800388 -->

@keivenchang Thanks for the information. I've addressed all the comments, we can also merge my PR.<br><br>BTW, Respond to your previous comment on parsing the etcd_path, we don't need to implement that. since `fn create_metric<T: PrometheusMetric, R: MetricsRegistry + ?Sized>( registry: &R,` already do that for us. So I think my PR is good to go.

keivenchang committed:
feat: implement etcd metrics initialization in distributed runtime

- Add etcd metrics initialization in DistributedRuntime::new()
- Use take() pattern to avoid borrowing conflicts when calling init_metrics
- Add comprehensive etcd_metrics_test to validate metrics functionality
- Add debug output in init_metrics to show existing instances
- Add documentation for kv_get_prefix method
- Fix doctest in etcd.rs to use no_run and proper async example

The implementation properly handles the circular reference between
DistributedRuntime and etcd::Client by using a temporary mutable
reference and cloning for the init_metrics call.

keivenchang reviewed (changes requested):
Getting closer, but still some work to do. 

1) In the future, please submit your changes linearly (no squash/force commit). Rebase when you absolutely need to, but avoided when possible.
1a) The Dynamo Onboarding guide 5) g) says this: "On your branch, you can keep pushing changes/merge during the review. We ask you to please avoid squashing+force committing, because reviewers want to see incremental change history. Also, note that when the code is finally merged to main, everything will be squashed anyways."

2) Test coverage on your changes is a must. Please add tests to all your code next time. The entry-point of a code-reviewer is/are your tests, where you set up clear input and clear expected output.

3) Please take a look at the changes I made for you. 
3a) I also added an integration test, which you can run (per the comments). Please run it to understand the test.
3b) Please look at the comments on kv_get_prefix to see what it does. 
3c) Please look at the note about the use of `mut` types.

keivenchang committed:
style: format debug print statements in etcd init_metrics

keivenchang committed:
chore: update Cargo.lock dependencies

keivenchang committed:
refactor: move etcd_metrics_test to transports/etcd.rs and fix integration test

tzulingk commented on lib/runtime/src/transports/etcd.rs:
```diff
@@ -382,6 +414,24 @@ impl Client {
+    /// # use dynamo_runtime::transports::etcd::Client;
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let client: Client = todo!(); // Replace with actual client creation
+    /// let instances = client.kv_get_prefix("instances/dynamo/backend/").await?;
```
I could only find keys starts with "instances" and "mdc" (LocalModel), could you point me to where we create keys with "storage" as prefix?
